module-system/module-implementations/sov-nft-module/src/offchain.rs

```diff
@@ -0,0 +1,142 @@
+#[cfg(feature = "offchain")]
+use postgres::NoTls;
+use sov_modules_macros::offchain;
+
+#[cfg(feature = "offchain")]
+use crate::utils::get_collection_address;
+#[cfg(feature = "offchain")]
+use crate::CollectionAddress;
+use crate::{Collection, Nft, OwnerAddress};
+
+/// Syncs a collection to the corresponding table "collections" in postgres
+#[offchain]
+pub fn update_collection<C: sov_modules_api::Context>(collection: &Collection<C>) {
+    // Extract the necessary metadata from the collection
+    let collection_name = collection.get_name();
+    let creator_address = collection.get_creator();
+    let frozen = collection.is_frozen();
+    let metadata_url = collection.get_collection_uri();
+    let supply = collection.get_supply();
+    let collection_address: CollectionAddress<C> =
+        get_collection_address(collection_name, creator_address.as_ref());
+    let collection_address_str = collection_address.to_string();
+    let creator_address_str = creator_address.to_string();
+    // postgres insert
+    tokio::task::block_in_place(|| {
+        if let Ok(conn_string) = std::env::var("POSTGRES_CONNECTION_STRING") {
+            match postgres::Client::connect(&conn_string, NoTls) {
+                Ok(mut client) => {
+                    let result = client.execute(
+                        "INSERT INTO collections (\
+                    collection_address, collection_name, creator_address,\
+                    frozen, metadata_url, supply)\
+                    VALUES ($1, $2, $3, $4, $5, $6)\
+                    ON CONFLICT (collection_address)\
+                    DO UPDATE SET collection_name = EXCLUDED.collection_name,\
+                                  creator_address = EXCLUDED.creator_address,\
+                                  frozen = EXCLUDED.frozen,\
+                                  metadata_url = EXCLUDED.metadata_url,\
+                                  supply = EXCLUDED.supply",
+                        &[
+                            &collection_address_str,
+                            &collection_name,
+                            &creator_address_str,
+                            &frozen,
+                            &metadata_url,
+                            &(supply as i64),
+                        ],
+                    );
+                    if let Err(e) = result {
+                        tracing::error!("Failed to execute query: {}", e);
+                    }
+                }
+                Err(e) => {
+                    tracing::error!("Failed to connect to the database: {}", e);
+                }
+            }
+        } else {
+            tracing::error!("Environment variable POSTGRES_CONNECTION_STRING is not set");
+        }
+    })
+}
+
+/// Syncs an NFT to the corresponding table "nfts" in postgres
+/// Additionally, this function also has logic to track the counts of NFTs held by each user
+/// in each collection.
+#[offchain]
```

Comment:
Can this macro be used with `Module` attributes. I am thinking that the `AccesorryXX` should be always behind offchain. @neysofu wdt?

Reply:
I very much agree, I think accessory state should be zero-cost and only written to during native execution. The current model works but we need to integrate it more tightly with `offchain`. Tracked in https://github.com/Sovereign-Labs/sovereign-sdk/issues/944.

Reply:
Cool @dubbelosix feel free to resolve this comment. We will do is as #944.